### PR TITLE
add ability to incrementally add/remove nodes from bvh

### PR DESCRIPTION
**Pull request by dbenson24**
leaving an initial draft of this PR here. It looks like there might be a perf regression with when I originally wrote this so I'm going to take a look at that. (post optimizaiton intersection of the 120k benchmark is massively improved, intersection with sponza isn't doing as well)

### CODE REVIEW

**Comment by dbenson24** [comment]
```
// PR
test bvh::optimization::bench::bench_intersect_120k_after_optimize_00p   ... bench:         853 ns/iter (+/- 15)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_01p   ... bench:         920 ns/iter (+/- 25)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_10p   ... bench:       2,607 ns/iter (+/- 488)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_50p   ... bench:       2,941 ns/iter (+/- 572)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_00p     ... bench:         853 ns/iter (+/- 21)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_01p     ... bench:         914 ns/iter (+/- 17)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_10p     ... bench:       1,745 ns/iter (+/- 243)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_50p     ... bench:       1,985 ns/iter (+/- 356)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_00p ... bench:       1,366 ns/iter (+/- 42)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_01p ... bench:       3,098 ns/iter (+/- 139)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_10p ... bench:      15,347 ns/iter (+/- 1,580)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_50p ... bench:       8,568 ns/iter (+/- 1,122)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_00p   ... bench:       1,359 ns/iter (+/- 33)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_01p   ... bench:       1,496 ns/iter (+/- 59)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_10p   ... bench:       1,906 ns/iter (+/- 105)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_50p   ... bench:       2,472 ns/iter (+/- 193)
test bvh::optimization::bench::bench_optimize_bvh_120k_00p               ... bench:   1,064,110 ns/iter (+/- 6,733)
test bvh::optimization::bench::bench_optimize_bvh_120k_01p               ... bench:   2,357,540 ns/iter (+/- 366,656)
test bvh::optimization::bench::bench_optimize_bvh_120k_10p               ... bench:  17,270,170 ns/iter (+/- 2,828,226)
test bvh::optimization::bench::bench_optimize_bvh_120k_50p               ... bench:  72,724,610 ns/iter (+/- 7,787,654)
test bvh::optimization::bench::bench_randomize_120k_50p                  ... bench:   4,599,345 ns/iter (+/- 1,906,880)
// Base
test bvh::optimization::bench::bench_intersect_120k_after_optimize_00p   ... bench:         855 ns/iter (+/- 24)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_01p   ... bench:     143,784 ns/iter (+/- 12,246)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_10p   ... bench:   1,297,513 ns/iter (+/- 622,827)
test bvh::optimization::bench::bench_intersect_120k_after_optimize_50p   ... bench:   2,366,247 ns/iter (+/- 1,195,041)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_00p     ... bench:         851 ns/iter (+/- 15)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_01p     ... bench:         919 ns/iter (+/- 18)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_10p     ... bench:       1,772 ns/iter (+/- 251)
test bvh::optimization::bench::bench_intersect_120k_with_rebuild_50p     ... bench:       2,058 ns/iter (+/- 434)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_00p ... bench:       1,367 ns/iter (+/- 50)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_01p ... bench:       2,873 ns/iter (+/- 197)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_10p ... bench:       4,147 ns/iter (+/- 492)
test bvh::optimization::bench::bench_intersect_sponza_after_optimize_50p ... bench:       5,918 ns/iter (+/- 695)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_00p   ... bench:       1,363 ns/iter (+/- 51)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_01p   ... bench:       1,498 ns/iter (+/- 67)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_10p   ... bench:       1,898 ns/iter (+/- 133)
test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_50p   ... bench:       2,484 ns/iter (+/- 216)
test bvh::optimization::bench::bench_optimize_bvh_120k_00p               ... bench:     984,960 ns/iter (+/- 83,377)
test bvh::optimization::bench::bench_optimize_bvh_120k_01p               ... bench:   2,079,140 ns/iter (+/- 245,858)
test bvh::optimization::bench::bench_optimize_bvh_120k_10p               ... bench:  12,014,960 ns/iter (+/- 1,541,641)
test bvh::optimization::bench::bench_optimize_bvh_120k_50p               ... bench:  53,112,990 ns/iter (+/- 4,470,184)
test bvh::optimization::bench::bench_randomize_120k_50p                  ... bench:   4,634,985 ns/iter (+/- 2,040,404)
```

**Comment by dbenson24** [comment]
Just gonna jot down my general arguments for this method over the current optimize.

The ability to dynamically add/remove individual nodes is a pretty major feature that you don't get from the other rust bvh (Parry's QBvh). This approach also eliminates the need to store the depth of each node so the nodes are smaller in memory (not massively smaller but it's still something). Keeping depth actually dramatically slows this approach down because it needs to update the depths during each operation.

All of this assumes that this approach also gets better/similar performance to the current one. Once I add the parallel rebuild, optimize/add/remove should be most relevant for changes of less than ~10% of the Bvh due to lack of parallelism.

There are still some optimizations left on the table. Right now add/remove doesn't leave the bvh in a contiguous state in memory. I want to test adding a relayout operation that puts all the nodes back in order. There's probably more operations

**Comment by svenstaro** [comment]
These are some incredible numbers! I might not get to review this but I hope @marstaik could take a look.

File: src/bvh/optimization.rs [hunk 1]
```diff
@@ -525,8 +414,356 @@ where
...
+                let node_index = shapes[deleted_shape_index].bh_node_index();
+                if let Some(index) = self.nodes[node_index].shape_index_mut() {
+                    *index = deleted_shape_index
+                }
+            }
+        }
+    }
+
+    /// Fixes bvh
+    pub fn optimize<'a, Shape: BHShape<T, D>>(
```
Comment (marstaik) [inline src/bvh/optimization.rs:668]: The name of this function doesn't match what it does anymore does it?
This is more of a `update_shapes()` call, as it removes and adds the nodes passed in. It doesn't actually perform any optimization that would make the tree more balanced.

File: src/bvh/optimization.rs [hunk 2]
```diff
@@ -525,8 +414,356 @@ where
...
+    ) {
+        for i in refit_shape_indices {
+            self.remove_node(shapes, *i, false);
+        }
+        for i in refit_shape_indices {
+            self.add_node(shapes, *i);
+        }
+    }
+
+    fn fix_aabbs_ascending<Shape: BHShape<T, D>>(&mut self, shapes: &mut [Shape], node_index: usize) {
```
Comment (marstaik) [inline src/bvh/optimization.rs:681]: What does `fix_aabbs` mean? Perhaps a different function name and some comments?
Reading this function tells me that this is more `recalculate_parent_aabbs_ascending` implying that a node  has changed and we want to traverse update and verify the parent aabs still contain the child

File: src/bvh/optimization.rs [hunk 3]
```diff
@@ -543,19 +780,20 @@ mod tests {
     #[test]
     /// Tests if [`Bvh::optimize()`] does not modify a fresh [`Bvh`].
     fn test_optimizing_new_bvh() {
-        let (shapes, mut bvh) = build_some_bh::<TBvh3>();
+        let (mut shapes, mut bvh) = build_some_bh::<TBvh3>();
```
Comment (marstaik) [inline src/bvh/optimization.rs:783]: maybe this should be called test_updating_shapes_new_bvh()

**Review by marstaik — CHANGES_REQUESTED** [review]
Really amazing job, I tried to be as thorough as I could.

A couple of notes

1.  The remove function leaves dead nodes in the vector, it would be better to `Tombstone` them (we can introduce a new enum value in BvhNode. This could be handled by an *actual* optimize pass in the future (more on that later).
2. (This isn't just related to your PR) optimize has a bad name IMO. We are not performing any optimization to the BVH, only removing shapes that have changed and re-adding them.

Reviewing your changes has given me some ideas for future performance increases (things you do not have to do in this PR):

- An actual optimization step, that removes `TombStone`'d Nodes and potentially re-balances the tree, and perhaps resort the hierarchy to a depth-first traversal. I believe DFS oriented setup could potentially make traversals faster.
- Making mutability of shapes occur via a handle of some kind, which when all changes to shapes are made can automatically optimize the tree when it is dropped.

**Comment by svenstaro** [comment]
@dbenson24 lil ping :)

**Comment by svenstaro** [comment]
@dbenson24 are you interested in continuing this?

**Comment by svenstaro** [comment]
Oh shit, congratz! Take all the time you need. :)

File: src/bvh/optimization.rs (hunk 2 above)
Comment (dbenson24) [reply src/bvh/optimization.rs]: this was all no longer needed code.

File: src/bvh/optimization.rs [hunk 4]
```diff
@@ -525,8 +414,356 @@ where
...
+
+                    // compared SA of the options
+                    let min_send = if send_left < send_right {
+                        send_left
+                    } else {
+                         send_right
+                    };
+                    // merge is more expensive only do when it's significantly better
+
+                    if merged < min_send * T::from_i8(3).unwrap() / T::from_i8(10).unwrap() {
```
[the remark this answers is not in the document]
Comment (dbenson24) [reply src/bvh/optimization.rs:128]: yes

**Comment by svenstaro** [comment]
Could you also write a CHANGELOG entry for changes that people should be aware of?

**Comment by dbenson24** [comment]
Can we do the changelog with my next PR and hold off on release until after that? I'll have it up shortly after this one is submitted

**Comment by svenstaro** [comment]
Oh ok, great.

**Comment by svenstaro** [comment]
@marstaik feel like taking another look at this one? It looks to me like all your last comments were addressed.

**Comment by marstaik** [comment]
@svenstaro Ill take a look soon, I was on a very long vacation :) I am back now.

**Comment by svenstaro** [comment]
@marstaik Friendly ping :)

File: src/bvh/optimization.rs [hunk 5]
```diff
@@ -525,8 +414,356 @@ where
...
+                        *self.nodes[child_l_index].parent_mut() = r_index;
+
+                        self.nodes[i] = BvhNode::Node {
+                            child_l_aabb: shape_aabb,
+                            child_l_index: l_index,
+                            child_r_aabb: merged_aabb,
+                            child_r_index: r_index,
+                            parent_index,
+                        };
+                        //self.fix_depth(l_index, depth + 1);
```
[the remark this answers is not in the document]
Comment (marstaik) [reply src/bvh/optimization.rs:156]: commented code still here

File: src/bvh/optimization.rs [hunk 6]
```diff
@@ -525,8 +414,356 @@ where
...
+            let gp_index = self.nodes[parent_index].parent();
+            //println!("{}->{}->{}", gp_index, parent_index, dead_node_index);
+
+            let sibling_index = if self.node_is_left_child(dead_node_index) {
+                self.nodes[parent_index].child_r()
+            } else {
+                self.nodes[parent_index].child_l()
+            };
+
+            // TODO: fix potential issue leaving empty spot in self.nodes
```
[the remark this answers is not in the document]
Comment (marstaik) [reply src/bvh/optimization.rs:270]: bump

**Review by marstaik — CHANGES_REQUESTED** [review]
Sorry, my life has also recently been hectic (out of country, new job in the last 6mo, etc) but should be a bit more stable for me to work on code now.

There are still some small code quality/renaming/ commented out functions that should be removed as they might cause confusion.

If you don't want to deal with the TODO related to empty nodes/tombstone-ing, maybe I can do it in a separate PR.

**Comment by svenstaro** [comment]
@marstaik Happy to have you back! Should we maybe just merge this and you apply your changes on top? It might be the more efficient approach.

**Comment by dbenson24** [comment]
No need, I think that is pretty much everything except for the tombstoning. I definitely don't want to do that in this PR. I want to get to the multithreading PR

**Comment by svenstaro** [comment]
@dbenson24 Great to see you're still around as well! :) Would be great if we could get this PR in soon.

**Comment by dbenson24** [comment]
I'm pretty sure this is ready to merge

**Comment by svenstaro** [comment]
Some more clippy stuff to fix :)

**Comment by svenstaro** [comment]
Weird, did we just spot a flaky test?

**Comment by svenstaro** [comment]
@marstaik What do you think?

**Comment by svenstaro** [comment]
@marstaik friendly ping :)

**Comment by svenstaro** [comment]
If I hear nothing for a week, I'll just merge this.

**Comment by svenstaro** [comment]
Finally merging this, thanks for you work @dbenson24!

**Comment by dbenson24** [comment]
I will get the parallelization PR up soon

**Pull request merged**